modes.go

```diff
@@ -343,48 +389,79 @@ func DoBatchedWrites(session *gocql.Session, threadResult *results.TestThreadRes
 		batchSize := 0
 
 		currentPk := workload.NextPartitionKey()
+		cks := []int64{}
+		values_sizes := []int{}
 		for !workload.IsPartitionDone() && atomic.LoadUint32(&stopAll) == 0 && batchSize < rowsPerRequest {
 			ck := workload.NextClusteringKey()
+			cks = append(cks, ck)
 			batchSize++
+
 			value := GenerateData(currentPk, ck, clusteringRowSizeDist.Generate())
+			values_sizes = append(values_sizes, len(value))
 			batch.Query(request, currentPk, ck, value)
 		}
 
-		requestStart := time.Now()
-		err := session.ExecuteBatch(batch)
-		requestEnd := time.Now()
-		if err != nil {
-			return err, time.Duration(0)
+		query_str := "BATCH >>>" + request + "; ck = " + fmt.Sprint(cks) + "; v_sizes = " + fmt.Sprint(values_sizes)
+		query_str = strings.Replace(query_str, "(?", "("+strconv.FormatInt(currentPk, 10), 1)
+		current_attempts := 0
+		for {
+			requestStart := time.Now()
+			err := session.ExecuteBatch(batch)
+			requestEnd := time.Now()
+
+			if err == nil {
+				rb.IncOps()
+				rb.AddRows(batchSize)
+				latency := requestEnd.Sub(requestStart)
+				return nil, latency
+			}
+			if retryHandler == "sb" {
+				err = handle_sb_retry_error(query_str, err, current_attempts)
+			}
+			if err != nil {
+				return err, time.Duration(0)
+			}
+			current_attempts++
 		}
-
-		rb.IncOps()
-		rb.AddRows(batchSize)
-
-		latency := requestEnd.Sub(requestStart)
-		return nil, latency
 	})
 }
 
 func DoCounterUpdates(session *gocql.Session, threadResult *results.TestThreadResult, workload WorkloadGenerator, rateLimiter RateLimiter) {
 	RunTest(threadResult, workload, rateLimiter, func(rb *results.TestThreadResult) (error, time.Duration) {
-		query := session.Query("UPDATE " + keyspaceName + "." + counterTableName +
-			" SET c1 = c1 + ?, c2 = c2 + ?, c3 = c3 + ?, c4 = c4 + ?, c5 = c5 + ? WHERE pk = ? AND ck = ?")
+		request := "UPDATE " + keyspaceName + "." + counterTableName +
+			" SET c1 = c1 + ?, c2 = c2 + ?, c3 = c3 + ?, c4 = c4 + ?, c5 = c5 + ? WHERE pk = ? AND ck = ?"
+		query := session.Query(request)
 		pk := workload.NextPartitionKey()
 		ck := workload.NextClusteringKey()
+		current_attempts := 0
+		query_str := request
+		query_str = strings.Replace(query_str, "+ ?", "+ "+strconv.FormatInt(ck, 10), 1)
+		query_str = strings.Replace(query_str, "+ ?", "+ "+strconv.FormatInt(ck+1, 10), 1)
+		query_str = strings.Replace(query_str, "+ ?", "+ "+strconv.FormatInt(ck+2, 10), 1)
+		query_str = strings.Replace(query_str, "+ ?", "+ "+strconv.FormatInt(ck+3, 10), 1)
+		query_str = strings.Replace(query_str, "+ ?", "+ "+strconv.FormatInt(ck+4, 10), 1)
+		query_str = strings.Replace(query_str, "pk = ?", "pk = "+strconv.FormatInt(pk, 10), 1)
+		query_str = strings.Replace(query_str, "ck = ?", "ck = "+strconv.FormatInt(ck, 10), 1)
```

Comment:
That's a lot of code just to format the query, and something similar is being done in multiple places. Do we need to substitute the question marks with the actual variables? Maybe the error generated in `handle_sb_retry_error` could just contain the query string with question marks, and - next to it - the query values printed with a default formatter.

```go
queryArgs := []interface{}{ck, ck+1, ck+2, ck+3, ck+4, pk, ck}
query.Bind(queryArgs...)

// ...

if retryHandler == "sb" {
    err = handle_sb_retry_error(query_str, queryArgs, err, current_attempts)
}

// ...

func handle_sb_retry_error(query_str string, args []interface{}, err error, current_attempts int) error {
    errorDesc := fmt.Sprintf("query: %v, args: %v || ERROR: ...", query_str, args, ...)
    // ...
}
```

Reply:
Actually it may be even much more simpler -> `query.String()`.
Thanks for pointing out at it.

Will update it.